Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2245,18 +2245,23 @@ def visit_CFuncDefNode(self, node):
             return node
         # transform @property decorators on ctypedef class functions
         for decorator_node in node.decorators[::-1]:
-            decorator = decorator_node.decorator
-            if decorator.is_name and decorator.name == 'property':
-                if len(node.decorators) > 1:
-                    return self._reject_decorated_property(node, decorator_node)
-                # Mark the node as a cgetter
-                node.type.is_cgetter = True
-                # Add a func_cname to be output instead of the attribute
-                node.entry.func_cname = node.body.stats[0].value.function.name
-                # done - remove the decorator node
+            _node = self.analyse_decorator(node, decorator_node.decorator)
+            if _node:
```

Comment:
Hmm, this refactoring doesn't look helpful. The return value means different things inside of the `analyse_decorator()` method, doesn't have an explanatory name here, and is basically ignored. That's rather obscure code.
Also, `analyse_decorator()` essentially consists of two code paths, but I have to look through the longest path in order to understand the shorter one (the one that says "nothing to do").

So, first of all, I think this whole transformation can be merged into the `DecoratorTransform`. Then, the `visit_DefNode()` method over there should share some code with `visit_CFuncDefNode()`. They will be doing different things in the end, but the search loop, conditions and error handling are the same, so they should only be implemented once. I propose to write a shared helper method that searches an `@property` (or `@setter` etc.) decorator, and, if found (and it's the only one), returns it, so that the two visitor methods can deal with it in their own ways. For all errors, it should just call `error()`. Failing to find a property decorator, returning `None` is a good enough signal to the caller.

Note that `_reject_decorated_property()` actually raises a `CompileError` in the `error()` call, so it does not actually return (I admit that that's not obvious either…).

Reply:
We reached the conclusion that an extra pass was needed in one of the previous iterations of this PR, since I need to mark the `node.type` attribute (which is an `Entry`), and that is calculated in the `AnalyseDeclarationsTransform` pass, which happens after the `DecoratorTransform` pass.

Reply:
Sorry, my bad. The usual problem of long-living PRs/branches/designs/…